---
title: "RNA-seq of SLC25A1 KD in human airway stem cells"
author: "Michael S Chimenti"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction

## Project Summary 
SLC25A1 encodes mitochondrial citrate carrier, which exports citrate from mitochondria to nuclear-cytosolic space, where it can be utilized to generate acetyl CoA for histone acetylation. SLC25A1 knockdown in airway stem cells results in less ciliated cell and more goblet cell differentiation.

Hypothesis: Citrate export vis SLC25A1 promotes ciliated cell differentiation by opening up regulatory regions of genes critical for ciliated cell differentiation

Aim 1: Identify genes differentially expressed between control and SLC25A1 KD (In particular, genes associated with ciliated cell and goblet cell differentiation)

Aim 2: Genome-wide correlation of gene expression and chromatin accessibility changes in control and SLC25A1 KD cells (using nucleosome positions and chromatic openness)

Aim 3: Identify previously unknown regulatory elements for airway stem cell differentiation

## Methods 

To be filled in later.  

## Library imports 
```{r, imports, warnings=FALSE, message=FALSE, eval=TRUE}
#negative binomial GLM and related
library('DESeq2')
library('readr')

#Exploratory analysis
library('tidyverse')
library('pcaExplorer')
library('ggrepel')

#pathway and gene clusters
#library('DEGreport')

#Heatmap 
library('pheatmap')
library('RColorBrewer')
library('viridis')

#Plotting
library(patchwork)

library('here')
```

## Function definitions 

```{r, function_defs}
## Volcano Plot
do_vol_plot <- function(df, sig=0.05, fc=0.5, size = 4){
  
  df_sig<- df %>% filter(padj < sig)
  
  ## genes with labels -- signficant,decreased expression and start with IGF
  
  df_label<- df %>% 
              filter(padj < sig, abs(log2FoldChange) > fc)
              #mutate(short_id = stringr::str_split(gene_id, "000000") %>% map_chr(.,2))
  
  
  ## plot
  
  p <- ggplot(df, aes(log2FoldChange, -log10(padj)))  +
      geom_point(size=0.8, color="black", alpha=.8) + 
      geom_point(size=0.8, data=df_sig, aes(log2FoldChange, -log10(padj)), colour="red") +
       geom_text_repel(size= size, 
                    max.overlaps = 25,
                    colour="black",
                    segment.size=0.1, 
                    nudge_x=0.06, 
                    nudge_y=0.06, 
                    data=df_label,
                    aes(log2FoldChange, -log10(padj), label=gene_name), 
                    max.iter= 200, 
                    point.padding = 0.15, 
                    segment.alpha = 1, 
                    box.padding=.15,
                    min.segment.length = unit(0.15, 'lines'),size=2.5) +
    theme(
      axis.text.x = element_text(size=12),
      axis.text.y = element_text(size=12),
      axis.title.x = element_text(size=14, margin = margin(t = 10, r = 0, b = 10, l = 0)),
      axis.title.y = element_text(size=14, margin = margin(t = 0, r = 10, b = 0, l = 10)),
      plot.margin =unit(c(.5,.5,.5,.5),"cm"),
      plot.title = element_text(size = 11)
    )
  
  return (p)
}
```


```{r}

meta <- readr::read_csv(here("data","sample_meta.csv")) %>% 
  mutate(across(.cols=-samplename, as.factor))

dat <- readr::read_delim(here("data", "salmon.merged.gene_counts_length_scaled.tsv"), delim = '\t', col_names = TRUE) %>%
  mutate(across(.cols=-c(gene_id,gene_name), as.integer))

meta <- meta %>% mutate(group=paste0(condition,"_",day)) %>% as.data.frame()
rownames(meta) <- meta$samplename
anno <- dat[1:2]
```

## Create DESEQ2 object

We know that the donor and prep batch can create problems, so we account for them in the model below: 

```{r, DESeq2, class.source = 'fold-show'}

dat <- dat[c(13,24,26,27,28,29,30,31,32,3,4,5,6,7,8,9,10,11,12,14,15,16,17,18,19,20,21,22,23,25)]
assertthat::are_equal(colnames(dat), meta$samplename)
dat <- as.matrix(dat)
rownames(dat) <- anno$gene_name

dds <- DESeqDataSetFromMatrix(countData = dat,
                              colData = meta,
                              design = ~ batch + donor + group)

dds <- dds[ rowSums(counts(dds)) > 10, ]
dds <- DESeq(dds)
```
## PCA analysis 

The PCA clearly shows that knockout does have an effect on expression, but the donor batch effect is significant as well.  PC3 shows this clearly.  

```{r}
vst.dds <- vst(dds)
pcaExplorer::pcaplot(vst.dds, intgroup="condition", ellipse = FALSE, ntop=400, text_labels = FALSE, title="PCA Plot: Control vs. Knockdown")
pcaExplorer::pcaplot(vst.dds, intgroup="day", ellipse = FALSE, ntop=400, text_labels = FALSE, title="PCA Plot: Day")
pcaExplorer::pcaplot(vst.dds, intgroup="batch", ellipse=FALSE, ntop=400, text_labels = FALSE, title="PCA Plot: Batch")
pcaExplorer::pcaplot(vst.dds, intgroup="donor", ellipse=FALSE, ntop=400, text_labels = FALSE, title="PCA Plot: Donor")
pcaExplorer::pcaplot(vst.dds, intgroup="group", ellipse=FALSE, ntop=400, text_labels = FALSE, title="PCA Plot: Grouped Day and KO")

pcaExplorer::pcaplot(vst.dds, intgroup = "donor", ellipse=FALSE, ntop=400, text_labels=FALSE, pcX = 2, pcY = 3, title = "PCA Plot PC2 and PC3: Batch effect with Donor")
pcaExplorer::pcaplot(vst.dds, intgroup = "group", ellipse=FALSE, ntop=400, text_labels=FALSE, pcX = 2, pcY = 3, title = "PCA Plot PC2 and PC3: Grouped Day and KO")
```
Here, we're looking at the association of the PCs and the factors in the experiment.  PC3 captures donor and batch effects.

```{r}
pcaobj_dds <- prcomp(t(assay(vst.dds)))
pcascree(pcaobj_dds, type='pev')
res_dds <- correlatePCs(pcaobj_dds, colData(dds))
plotPCcorrs(res_dds, pc = 1)
plotPCcorrs(res_dds, pc = 2)
plotPCcorrs(res_dds, pc = 3)
```

## Sample to sample heatmap 

This is a sample-to-sample distance heatmap.  Here again we can see the donor batch effect, particularly for Donor "A".  

```{r}

sampleDists <- dist(t(assay(vst.dds)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vst.dds$samplename
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
anno_col <- as.data.frame(colData(vst.dds)[,4:6], row.names = vst.dds$sample)
pheatmap(sampleDistMatrix,
         #annotation_col = anno_col,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         colors = colors,
         annotation_row = anno_col)
```         
         
## Analyze DE changes

### Day 0 

```{r}
res_KO_Day0 <- as.data.frame(results(dds,contrast = c('group','ko_d0','cont_d0')))
## Table of top DE genes 
res_KO_Day0_sig <- res_KO_Day0 %>% 
  arrange(padj) %>% 
  filter(padj < 0.00001) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_KO_Day0_sig
```

```{r, volPlot, fig.width=9, fig.height=9}
df <- res_KO_Day0 %>% mutate(gene_name = rownames(.)) %>% filter(padj < 0.0001) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-15, fc = 1, size = 2)
p
```
### Knockout Sanity Check 

```{r}
plotCounts(dds, gene = "SLC25A1", intgroup = "group")
plotCounts(dds, gene = "DYNLRB1", intgroup = "group")
plotCounts(dds, gene = "DBNDD2", intgroup = "group")
plotCounts(dds, gene = "TMED4", intgroup = "group")

```


### Day 1
```{r}
res_KO_Day1 <- as.data.frame(results(dds,contrast = c('group','ko_d1','cont_d1')))
## Table of top DE genes 
res_KO_Day1_sig <- res_KO_Day1 %>% 
  arrange(padj) %>% 
  filter(padj < 0.00001) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_KO_Day1_sig
```

```{r}
df <- res_KO_Day1 %>% mutate(gene_name = rownames(.)) %>% filter(padj < 0.0001) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-15, fc = 1, size = 2)
p
```

### Day 2
```{r}
res_KO_Day2 <- as.data.frame(results(dds,contrast = c('group','ko_d2','cont_d2')))
## Table of top DE genes 
res_KO_Day2_sig <- res_KO_Day2 %>% 
  arrange(padj) %>% 
  filter(padj < 0.00001) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_KO_Day2_sig
```

```{r, fig.dim=c(9,9)}
df <- res_KO_Day2 %>% mutate(gene_name = rownames(.)) %>% filter(padj < 0.0001) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-15, fc = 1, size = 2)
p
```

### Control Day 2 vs Day 1

```{r}
res_cont_d2_d1 <- as.data.frame(results(dds,contrast = c('group','cont_d2','cont_d1')))
## Table of top DE genes 
res_cont_d2_d1_sig <- res_cont_d2_d1 %>% 
  arrange(padj) %>% 
  filter(padj < 0.001) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_cont_d2_d1_sig
```


```{r, fig.dim=c(9,9)}
df <- res_cont_d2_d1 %>% mutate(gene_name = rownames(.)) %>% filter(padj < 0.01) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-4, fc = 0.7, size = 2)
p
```

### Knockdown day 2 vs day 1

```{r}
res_ko_d2_d1 <- as.data.frame(results(dds,contrast = c('group','ko_d2','ko_d1')))
## Table of top DE genes 
res_ko_d2_d1_sig <- res_ko_d2_d1 %>% 
  arrange(padj) %>% 
  filter(padj < 0.001) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_ko_d2_d1_sig
```

```{r, fig.dim=c(9,9)}
df <- res_ko_d2_d1 %>% mutate(gene_name = rownames(.)) %>% filter(padj < 0.01) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-3, fc = 0.7, size = 2)
p
```

### Differential effect of knockout by day 

Here we are calculating a group of genes that behave differently across days of the timecourse depending on their genotype.  We do this by modeling a genotype (called "condition" here) and day interaction term, and using a likelihood ratio test to find genes that are better fit to that model with the interaction term than without. 



```{r}
dds2 <- DESeqDataSetFromMatrix(countData = dat,
                              colData = meta,
                              design = ~ batch + donor + day*condition)



dds2 <- dds2[ rowSums(counts(dds)) > 10, ]
dds2 <- DESeq(dds2, test = "LRT", reduced = ~ batch + donor + day + condition)
res_interact <- results(dds2)

res_interact_sig <- res_interact %>% 
  as.data.frame() %>%
  arrange(padj) %>% 
  filter(padj < 0.01) %>% 
  dplyr::select(c(padj, baseMean, log2FoldChange)) %>% 
  mutate(across(c(1:3), ~round(.x, digits=3)))
  
  
res_interact_sig
```

Top genes that change behavior with time in the Knockout group: 

```{r}
plotCounts(dds, gene = "LOXL2", intgroup = "group")
plotCounts(dds, gene = "COL17A1", intgroup = "group")
plotCounts(dds, gene = "PSEN2", intgroup = "group")
plotCounts(dds, gene = "S100A10", intgroup = "group")

```
These are genes that are significantly different across time with KO and WT status: 

```{r, fig.height=9, fig.width=9, dpi=300}
df <- res_interact %>% 
  as.data.frame() %>% 
  mutate(gene_name = rownames(.)) %>% filter(padj < 0.05) %>% filter(abs(log2FoldChange) < 10)
p <- do_vol_plot(df = df, sig = 1e-3, fc = 1.2, size = 2)
p
```
```{r, eval=FALSE}
write.csv(res_KO_Day0_sig, here("data","DEGs_KO_WT_Day0_sig_p1Em5.csv"))
write.csv(res_KO_Day1_sig, here("data","DEGs_KO_WT_Day1_sig_p1Em5.csv"))
write.csv(res_KO_Day2_sig, here("data","DEGs_KO_WT_Day2_sig_p1Em5.csv"))
write.csv(res_interact_sig, here("data","DEGs_Geno_Day_Interact_sig_p1Em2.csv"))

res_KO_Day0$gene_name <- rownames(res_KO_Day0)
write_delim(x=res_KO_Day0, file=here("data","res_KO_Day0_foriPath.tsv"), delim = '\t')

res_KO_Day1$gene_name <- rownames(res_KO_Day1)
write_delim(x=res_KO_Day1, file=here("data","res_KO_Day1_foriPath.tsv"), delim = '\t')

res_KO_Day2$gene_name <- rownames(res_KO_Day2)
write_delim(x=res_KO_Day2, file=here("data","res_KO_Day2_foriPath.tsv"), delim = '\t')

res_interact$gene_name <- rownames(res_interact)
write_delim(x=as.data.frame(res_interact), file=here("data","res_interact_foriPath.tsv"), delim = '\t')
```


```{r}
sessionInfo()

```




